packages/etcd/extra/etcd_discovery/etcd_discovery.py

```diff
@@ -0,0 +1,641 @@
+#!/usr/bin/env python
+
+import argparse
+import ipaddress
+import json
+import logging
+import os
+import random
+import socket
+import subprocess
+import sys
+from contextlib import contextmanager
+from typing import Dict, Generator, List, Optional, Union
+
+from kazoo.client import KazooClient
+from kazoo.exceptions import ConnectionLoss, LockTimeout, SessionExpiredError
+from kazoo.retry import KazooRetry
+from kazoo.security import make_digest_acl
+
+log = logging.getLogger(__name__)
+
+JsonTypeMembers = List[Dict[str, Union[str, int, List[str]]]]
+
+# The path of the ZNode used for locking.
+ZK_LOCK_PATH = "/etcd/locking"
+# The path of the ZNode containing the list of cluster members.
+ZK_NODES_PATH = "/etcd/nodes"
+# The id to use when contending for the ZK lock.
+LOCK_CONTENDER_ID = "{}:{}".format(socket.gethostname(), os.getpid())
+# The time in seconds to wait when attempting to acquire a lock.  Lock
+# acquisition between 5 ZooKeeper nodes is an operation on the order of
+# milliseconds.
+#
+# Furthermore, the operations performed while the lock is held are performed
+# once and never again. This means a process will only contend for the lock
+# once. As such, if lock aquisition fails due to some other process holding it,
+# the current process will crash and be restarted with one less contender for
+# the same lock. This means that the locking behaviour does converge and no
+# timeout-sensitive livelock can occur.
+#
+# We set the lock timeout to a couple of seconds instead of milliseconds to
+# account for variation in network latency between nodes in the cluster. The
+# current value has so far shown to be sufficient.
+ZK_LOCK_TIMEOUT = 5
+# Location of the detect IP address detection script
+DETECT_IP_SCRIPT = '/opt/mesosphere/bin/detect_ip'
+
+
+# In theory we could use the dcos_internal_utils.utils library, but in practice
+# the amount of deps it pulls in is huge compared to just a simple `execute
+# script and verify the output` function that we want
+def detect_ip() -> str:
+    machine_ip = subprocess.check_output(
+        [DETECT_IP_SCRIPT], stderr=subprocess.DEVNULL).decode('ascii').strip()
```

Comment:
Can we try to not hide any errors here? `stderr=subprocess.STDOUT` may do the trick.

Reply:
Done.